rustls-platform-verifier/src/verification/windows.rs

```diff
@@ -19,72 +19,84 @@
 
 use super::{log_server_cert, ALLOWED_EKUS};
-use crate::windows::{
-    c_void_from_ref, c_void_from_ref_mut, nonnull_from_const_ptr, ZeroedWithSize,
-};
 use once_cell::sync::OnceCell;
 use rustls::client::danger::{HandshakeSignatureValid, ServerCertVerifier};
 use rustls::crypto::{verify_tls12_signature, verify_tls13_signature, CryptoProvider};
 use rustls::pki_types;
-use rustls::{CertificateError, DigitallySignedStruct, Error as TlsError, SignatureScheme};
-use winapi::{
-    shared::{
-        minwindef::{FILETIME, TRUE},
-        ntdef::{LPSTR, VOID},
-        winerror::{
-            CERT_E_CN_NO_MATCH, CERT_E_EXPIRED, CERT_E_INVALID_NAME, CERT_E_UNTRUSTEDROOT,
-            CERT_E_WRONG_USAGE, CRYPT_E_REVOKED,
-        },
+use rustls::{
+    CertificateError, DigitallySignedStruct, Error as TlsError, Error::InvalidCertificate,
+    SignatureScheme,
+};
+use std::{
+    convert::TryInto,
+    mem::{self, MaybeUninit},
+    os::raw::c_void,
+    ptr::{self, NonNull},
+    sync::Arc,
+};
+use windows_sys::Win32::{
+    Foundation::{
+        BOOL, CERT_E_CN_NO_MATCH, CERT_E_EXPIRED, CERT_E_INVALID_NAME, CERT_E_UNTRUSTEDROOT,
+        CERT_E_WRONG_USAGE, CRYPT_E_REVOKED, FILETIME, TRUE,
     },
-    um::wincrypt::{
+    Security::Cryptography::{
         CertAddEncodedCertificateToStore, CertCloseStore, CertFreeCertificateChain,
         CertFreeCertificateChainEngine, CertFreeCertificateContext, CertGetCertificateChain,
         CertOpenStore, CertSetCertificateContextProperty, CertVerifyCertificateChainPolicy,
-        AUTHTYPE_SERVER, CERT_CHAIN_CACHE_END_CERT, CERT_CHAIN_CONTEXT, CERT_CHAIN_PARA,
+        HTTPSPolicyCallbackData, AUTHTYPE_SERVER, CERT_CHAIN_CACHE_END_CERT, CERT_CHAIN_CONTEXT,
         CERT_CHAIN_POLICY_IGNORE_ALL_REV_UNKNOWN_FLAGS, CERT_CHAIN_POLICY_PARA,
         CERT_CHAIN_POLICY_SSL, CERT_CHAIN_POLICY_STATUS,
         CERT_CHAIN_REVOCATION_ACCUMULATIVE_TIMEOUT, CERT_CHAIN_REVOCATION_CHECK_END_CERT,
         CERT_CONTEXT, CERT_OCSP_RESPONSE_PROP_ID, CERT_SET_PROPERTY_IGNORE_PERSIST_ERROR_FLAG,
         CERT_STORE_ADD_ALWAYS, CERT_STORE_DEFER_CLOSE_UNTIL_LAST_FREE_FLAG, CERT_STORE_PROV_MEMORY,
-        CERT_USAGE_MATCH, CRYPT_DATA_BLOB, CTL_USAGE, SSL_EXTRA_CERT_CHAIN_POLICY_PARA,
+        CERT_STRONG_SIGN_PARA, CERT_USAGE_MATCH, CRYPT_INTEGER_BLOB, CTL_USAGE,
         USAGE_MATCH_TYPE_AND, X509_ASN_ENCODING,
     },
 };
 
-use rustls::Error::InvalidCertificate;
-use std::{
-    convert::TryInto,
-    mem::{self, MaybeUninit},
-    ptr::{self, NonNull},
-    sync::Arc,
-};
+// XXX: The `windows-sys` definition for `CERT_CHAIN_PARA` is missing the optional fields.
+// See https://github.com/microsoft/win32metadata/commit/185c3ef016663ab8201878da7efab0a2cff0bcc7.
+//
+// However: by defining it ourselves we can keep better (hypothetical) OS backwards compat.
+// When its available in `windows-sys`, a compile-time size assertion can be added to help it stay in sync.
```

Comment:
The upstream commit only fixes the definition of the entire structure. I am not aware of an issue tracking the conditional definition of each field since, AFAIK, the Windows Metadata project itself isn't interested in backwards compatibility with old OSes.

Reply:
> I am not aware of an issue tracking the conditional definition of each field since, AFAIK, the Windows Metadata project itself isn't interested in backwards compatibility with old OSes.

In that case does it seem likely that a matching structure will ever be available upstream? If not, maybe it's simpler to just remove the comment that suggests it could be replaced by a compile-time assertion against the upstream definition.

Reply:
That sounds plenty reasonable to me. I edited the comment to match this 👍.

Reply:
Thanks!